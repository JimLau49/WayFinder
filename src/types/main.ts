export interface Location {
  latitude: number;
  longitude: number;
}

export interface Region extends Location {
  latitudeDelta: number;
  longitudeDelta: number;
}

export interface Building {
  id: BuildingId;
  campusId: CampusId;
  displayName: string;
  address: string;
  location: Location;
  departments: LinkItem[];
  services: LinkItem[];
  boundingBox: Location[];
}

export interface LinkItem {
  id: number;
  title: string;
  link: string;
}

export interface Campus {
  id: CampusId;
  displayName: string;
  description: string;
  region: Region;
  buildings: BuildingId[];
}

export interface MarkerLocation {
  id: string;
  displayName: string;
  location: Location;
}

export interface IndoorLocation extends MarkerLocation {
  buildingId: BuildingId;
  level: number;
}

export interface POI extends IndoorLocation {
  description: string;
  category: POICategory;
}

export interface Connector extends POI {
  type: ConnectorType;
  connections: Connection[];
}

export interface Connection {
  level: number;
  location: Location;
}

export interface BuildingFloor {
  id: number;
  buildingId: BuildingId;
  level: number;
  bounds: [Coordinate, Coordinate];
  image: any;
  travelNodes: TravelNode[];
}

export interface TravelNode {
  id: number;
  location: Location;
  children: number[];
}

export interface IndoorInformation {
  currentFloor: IndoorFloor;
  floors: IndoorFloor[];
}

export interface IndoorFloor {
  level: number;
  index: number;
}

export interface Range {
  min: number;
  max: number;
}

export type PQItem = { id: number; parent: number; g: number };

export type TravelEdge = [TravelNode, TravelNode];

export type Line = [Location, Location];

export type Coordinate = [number, number];

export enum CampusId {
  SGW = "SGW",
  Loyola = "Loyola"
}

export enum BuildingId {
  // SGW Campus
  B,
  CB,
  CI,
  CL,
  D,
  EN,
  ER,
  EV,
  FA,
  FB,
  FG,
  GA,
  GM,
  GN,
  GS,
  H,
  K,
  LB,
  LD,
  LS,
  M,
  MB,
  MI,
  MU,
  P,
  PR,
  Q,
  R,
  RR,
  S,
  SB,
  T,
  TD,
  V,
  VA,
  X,
  Z,
  // Loyola Campus
  AD,
  BB,
  BH,
  CC,
  CJ,
  DO,
  FC,
  GE,
  HA,
  HB,
  HC,
  HU,
  JR,
  PC,
  PS,
  PT,
  PY,
  QA,
  RA,
  RF,
  SC,
  SH,
  SI,
  SP,
  TA,
  VE,
  VL
}

export enum ZoomLevel {
  // Too far
  NONE,
  // Really zoomed out
  CAMPUS,
  // Medium
  OUTDOOR,
  // Close
  INDOOR
}

export enum POICategory {
  Classroom,
  Washroom,
  Restaurant,
  Association,
  Elevator,
  Escalator,
  Stairs,
  Exit,
  Service
}

export enum ConnectorType {
  Elevator,
  Stairs,
  Ramp,
  Escalator
}

export enum TravelState {
  NONE,
  PLANNING,
  TRAVELLING
}
