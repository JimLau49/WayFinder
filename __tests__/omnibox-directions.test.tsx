import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { Platform } from "react-native";
import { getOmniboxAutoCompleteHeight } from "../src/services/auto-complete-height.service";
import OmniboxDirections from "../src/components/search/omnibox-directions.component";
import { POI, BuildingId, POICategory, TravelState } from "../src/types/main";

const mockPOIs: POI[] = [
  {
    id: "c4541028-739d-4651-b9f5-dc4d54054807",
    displayName: "H803",
    description: "",
    location: {
      latitude: 45.497482,
      longitude: -73.579034
    },
    buildingId: BuildingId.H,
    level: 8,
    category: POICategory.Washroom
  },
  {
    id: "f5a7d95c-170b-48c9-aac3-e9faafa6033d",
    displayName: "H805",
    description: "",
    location: {
      latitude: 45.497253,
      longitude: -73.579251
    },
    buildingId: BuildingId.H,
    level: 8,
    category: POICategory.Washroom
  }
];

describe("OmniboxDirections component", () => {
  it("should match snapshot", () => {
    const tree = renderer
      .create(
        <OmniboxDirections
          startLocation={mockPOIs[0]}
          endLocation={mockPOIs[1]}
          currentLocation={null}
          setEndLocation={null}
          setStartLocation={null}
          setEndLocationFocused={null}
          endLocationFocused={null}
          setTravelState={null}
          updateSearchResults={null}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should run provided function when back arrow is pressed", () => {
    const mockSetEndLocation = jest.fn();
    const mockSetStartLocation = jest.fn();
    const mockSetEndLocationFocused = jest.fn();
    const mockSetTravelState = jest.fn();

    const wrapper = shallow(
      <OmniboxDirections
        setEndLocation={mockSetEndLocation}
        setStartLocation={mockSetStartLocation}
        setEndLocationFocused={mockSetEndLocationFocused}
        startLocation={mockPOIs[0]}
        setTravelState={mockSetTravelState}
        endLocation={mockPOIs[1]}
        currentLocation={null}
        updateSearchResults={null}
        endLocationFocused={null}
      />
    );

    const backArrow = wrapper.find({ testID: "backArrow" });
    backArrow.simulate("press");

    expect(mockSetEndLocation).toHaveBeenCalledTimes(1);
    expect(mockSetTravelState).toHaveBeenCalledTimes(1);
    expect(mockSetEndLocation.mock.calls[0][0]).toBe(null);
    expect(mockSetTravelState.mock.calls[0][0]).toBe(TravelState.NONE);
  });

  it("should call provided function when start location textInput text is changed", () => {
    const mockUpdateSearchResults = jest.fn();

    const wrapper = shallow(
      <OmniboxDirections
        startLocation={mockPOIs[0]}
        endLocation={mockPOIs[1]}
        currentLocation={null}
        endLocationFocused={null}
        setEndLocation={null}
        setStartLocation={null}
        setEndLocationFocused={null}
        setTravelState={null}
        updateSearchResults={mockUpdateSearchResults}
      />
    );

    const textInput = wrapper.find({ testID: "searchInputInitialLocation" });
    textInput.simulate("changeText", "H803");
    expect(mockUpdateSearchResults).toHaveBeenCalledTimes(1);
    expect(mockUpdateSearchResults.mock.calls[0][0]).toBe("H803");
  });

  it("should call the provided function on focus of start location TextInput", () => {
    const mockSetEndLocationFocused = jest.fn();

    const wrapper = shallow(
      <OmniboxDirections
        startLocation={mockPOIs[0]}
        endLocation={mockPOIs[1]}
        setEndLocationFocused={mockSetEndLocationFocused}
        currentLocation={null}
        setEndLocation={null}
        setStartLocation={null}
        updateSearchResults={null}
        endLocationFocused={null}
        setTravelState={null}
      />
    );

    const textInput = wrapper.find({ testID: "searchInputInitialLocation" });
    textInput.simulate("focus");
    expect(mockSetEndLocationFocused).toHaveBeenCalledTimes(1);
    expect(mockSetEndLocationFocused.mock.calls[0][0]).toBe(false);
  });

  it("should call a function when end location textInput text is changed", () => {
    const mockUpdateSearchResults = jest.fn();

    const wrapper = shallow(
      <OmniboxDirections
        startLocation={mockPOIs[0]}
        endLocation={mockPOIs[1]}
        currentLocation={null}
        setEndLocation={null}
        setStartLocation={null}
        setEndLocationFocused={null}
        endLocationFocused={null}
        setTravelState={null}
        updateSearchResults={mockUpdateSearchResults}
      />
    );

    const textInput = wrapper.find({
      testID: "searchInputDestinationLocation"
    });
    textInput.simulate("changeText", "H805");
    expect(mockUpdateSearchResults).toHaveBeenCalledTimes(1);
    expect(mockUpdateSearchResults.mock.calls[0][0]).toBe("H805");
  });

  it("should call the provided functin when end location textInput is focused", () => {
    const mockSetEndLocationFocused = jest.fn();

    const wrapper = shallow(
      <OmniboxDirections
        startLocation={mockPOIs[0]}
        endLocation={mockPOIs[1]}
        setEndLocationFocused={mockSetEndLocationFocused}
        currentLocation={null}
        setEndLocation={null}
        setStartLocation={null}
        endLocationFocused={null}
        setTravelState={null}
        updateSearchResults={null}
      />
    );

    const textInput = wrapper.find({
      testID: "searchInputDestinationLocation"
    });
    textInput.simulate("focus");
    expect(mockSetEndLocationFocused).toHaveBeenCalledTimes(1);
    expect(mockSetEndLocationFocused.mock.calls[0][0]).toBe(true);
  });

  it("should call setshowTimePicker() when pressed", () => {
    const mockSetShowTimePicker = jest.fn();

    const useStateSpy = jest.spyOn(React, "useState");
    // @ts-ignore
    useStateSpy.mockImplementation(init => [init, mockSetShowTimePicker]);

    const wrapper = shallow(
      <OmniboxDirections
        startLocation={mockPOIs[0]}
        endLocation={mockPOIs[1]}
        currentLocation={null}
        setEndLocation={null}
        setStartLocation={null}
        setEndLocationFocused={null}
        endLocationFocused={null}
        setTravelState={null}
        updateSearchResults={null}
      />
    );

    wrapper
      .find({ testID: "timePickerButton" })
      .props()
      .onPress();

    expect(mockSetShowTimePicker).toHaveBeenCalledTimes(1);
  });

  it("should get the correct height of the AutoComplete component", () => {
    let startSearchResultValues: any[];
    let startLocationInputValue: string;
    let endLocationSearchResults: any[];
    let showTimePicker: boolean;
    let endLocationInputValue: string;

    let autoComplete = getOmniboxAutoCompleteHeight(
      startSearchResultValues,
      startLocationInputValue,
      endLocationSearchResults,
      (showTimePicker = null),
      (endLocationInputValue = null)
    );
    expect(autoComplete).toBe(260);

    startSearchResultValues = ["1"];
    startLocationInputValue = "test";
    endLocationSearchResults = null;
    showTimePicker = null;
    endLocationInputValue = null;

    autoComplete = getOmniboxAutoCompleteHeight(
      startSearchResultValues,
      startLocationInputValue,
      endLocationSearchResults,
      (showTimePicker = null),
      (endLocationInputValue = null)
    );
    expect(autoComplete).toBe(310);

    startSearchResultValues = [];
    startLocationInputValue = "";
    endLocationSearchResults = ["1"];
    showTimePicker = null;
    endLocationInputValue = "test";

    autoComplete = getOmniboxAutoCompleteHeight(
      startSearchResultValues,
      startLocationInputValue,
      endLocationSearchResults,
      (showTimePicker = null),
      endLocationInputValue
    );
    expect(autoComplete).toBe(310);

    startSearchResultValues = [];
    startLocationInputValue = null;
    endLocationSearchResults = null;
    showTimePicker = true;
    endLocationInputValue = null;

    Platform.OS = "ios";

    autoComplete = getOmniboxAutoCompleteHeight(
      startSearchResultValues,
      startLocationInputValue,
      endLocationSearchResults,
      (showTimePicker = null),
      (endLocationInputValue = null)
    );
    expect(autoComplete).toBe(235);
  });
});
