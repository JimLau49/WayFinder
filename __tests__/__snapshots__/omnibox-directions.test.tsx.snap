// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OmniboxDirections component should match snapshot 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#F2F2F2",
          "height": 235,
          "position": "absolute",
          "top": 0,
          "width": 750,
          "zIndex": 1,
        },
        Object {
          "height": 260,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginLeft": 15,
          "marginRight": 15,
          "paddingBottom": 20,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="backArrow"
      >
        <Text />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginLeft": 36,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 80,
              "resizeMode": "contain",
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            selectTextOnFocus={true}
            style={
              Object {
                "borderColor": "#DEDEDE",
                "borderWidth": 1,
                "fontSize": 16,
                "height": 38,
                "marginBottom": 15,
                "marginLeft": 10,
                "paddingLeft": 10,
              }
            }
            testID="searchInputInitialLocation"
            underlineColorAndroid="transparent"
            value={null}
          />
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            rejectResponderTermination={true}
            selectTextOnFocus={true}
            style={
              Object {
                "borderColor": "#DEDEDE",
                "borderWidth": 1,
                "fontSize": 16,
                "height": 38,
                "marginBottom": 15,
                "marginLeft": 10,
                "paddingLeft": 10,
              }
            }
            testID="searchInputDestinationLocation"
            underlineColorAndroid="transparent"
            value="H805"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 0,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#DEDEDE",
              "borderWidth": 2,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "space-between",
              "marginLeft": 36,
              "opacity": 1,
              "paddingLeft": 15,
              "paddingRight": 15,
            }
          }
          testID="timePickerButton"
        >
          <Text
            style={
              Object {
                "color": "#717171",
              }
            }
          >
            DEPART NOW
          </Text>
          <Text />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "height": 26,
                "resizeMode": "contain",
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTSafeAreaView>,
  <View
    style={
      Object {
        "alignSelf": "center",
        "bottom": 40,
        "flex": 1,
        "height": 60,
        "position": "absolute",
        "right": 20,
        "width": 60,
        "zIndex": 1,
      }
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        style={
          Object {
            "height": 60,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#AA2B45",
              "borderColor": "rgba(0,0,0,0.2)",
              "borderRadius": 50,
              "borderWidth": 1,
              "flex": 1,
              "height": 60,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": 60,
            }
          }
        >
          <Text />
        </View>
      </View>
    </RNGestureHandlerButton>
  </View>,
]
`;
